File: М11_ClusterVMS/clustervms/cluster/epoch.py
"""Lesson 28 — the epoch by check-and-set, and the lease that fences.

The epoch is a fencing token: it must come from ONE issuer and it must
INCREASE. A Nomad Variable with cas is both. A variable lock is neither
(its ID is an opaque UUID). A database sequence reissues numbers after a
restore. The archive checks the token by construction: the epoch is in the
path, so a stale instance cannot name the live files.

The lease is what makes the stale instance STOP, on a monotonic clock:
    may write while  now − last_renewal < TTL − margin
Renewal is: read my own epoch Variable and find it still mine. A different
epoch there means a replacement was issued one; that is the moment the
zombie learns what it is, and node_epoch_conflicts is the counter.
"""
from __future__ import annotations

import time

from .variables import Conflict, Variables


def next_epoch(vars_: Variables, node: str, retries: int = 10) -> tuple[int, int]:
    """Issue the next epoch for `node`. Returns (epoch, ModifyIndex). Two
    callers racing get two DIFFERENT epochs, in order; the loser of the CAS
    re-reads and goes again. Nobody ever receives the same number."""
    path = f"nodes/{node}/epoch"
    for _ in range(retries):
        items, idx = vars_.get(path)
        current = int(items["epoch"]) if items else 0
        try:
            new_idx = vars_.put(path, {"epoch": current + 1}, cas=idx)
            return current + 1, new_idx
        except Conflict:
            continue
    raise RuntimeError(f"could not issue an epoch for {node} after {retries} conflicts")


def current_epoch(vars_: Variables, node: str) -> int:
    items, _ = vars_.get(f"nodes/{node}/epoch")
    return int(items["epoch"]) if items else 0


class Lease:
    """Held by the running instance. `renew()` is a read of the epoch
    Variable; `may_write()` is a purely local decision on a monotonic clock."""

    def __init__(self, vars_: Variables, node: str, epoch: int, ttl: float = 30.0,
                 margin: float = 5.0, clock=time.monotonic):
        self.vars, self.node, self.epoch = vars_, node, epoch
        self.ttl, self.margin, self.clock = ttl, margin, clock
        self.last_renewal = clock()
        self.fenced = False
        self.conflicts = 0                         # node_epoch_conflicts

    def renew(self) -> bool:
        """True if the lease still holds. False (and fenced) if the cluster
        issued a newer epoch — or if the read failed and the TTL ran out."""
        if self.fenced:
            return False
        try:
            live = current_epoch(self.vars, self.node)
        except Exception:                          # noqa: BLE001 — the cluster is unreachable; keep going until TTL - margin
            return self.may_write()
        if live != self.epoch:
            self.fenced = True
            self.conflicts += 1
            return False
        self.last_renewal = self.clock()
        return True

    def may_write(self) -> bool:
        if self.fenced:
            return False
        return (self.clock() - self.last_renewal) < (self.ttl - self.margin)

    def seconds_left(self) -> float:
        return max(0.0, (self.ttl - self.margin) - (self.clock() - self.last_renewal))

File: М11_ClusterVMS/clustervms-go/cluster/epoch.go
package cluster

// Lesson 3 — the epoch by check-and-set, and the lease that fences.
//
// The epoch is a fencing token: it must come from ONE issuer and it must
// INCREASE. A Nomad Variable with cas is both. A variable lock is neither
// (its ID is an opaque UUID). A database sequence reissues numbers after a
// restore. The archive checks the token by construction: the epoch is in
// the path, so a stale instance cannot name the live files.
//
// The lease is what makes the stale instance STOP, on a monotonic clock:
//
//	may write while  now − last_renewal < TTL − margin
//
// Renewal is: read my own epoch Variable and find it still mine. A different
// epoch there means a replacement was issued one; that is the moment the
// zombie learns what it is, and node_epoch_conflicts is the counter.

import (
	"errors"
	"fmt"
	"strconv"
	"time"
)

func epochPath(node string) string { return "nodes/" + node + "/epoch" }

// NextEpoch issues the next epoch for node. Two callers racing get two
// DIFFERENT epochs, in order; the loser of the CAS re-reads and goes again.
// Nobody ever receives the same number.
func NextEpoch(v Variables, node string) (epoch, index int64, err error) {
	const retries = 10
	for i := 0; i < retries; i++ {
		items, idx, err := v.Get(epochPath(node))
		if err != nil {
			return 0, 0, err
		}
		current := atoi(items["epoch"])
		newIdx, err := v.Put(epochPath(node), Items{"epoch": strconv.FormatInt(current+1, 10)}, idx)
		if errors.Is(err, ErrConflict) {
			continue
		}
		if err != nil {
			return 0, 0, err
		}
		return current + 1, newIdx, nil
	}
	return 0, 0, fmt.Errorf("could not issue an epoch for %s after %d conflicts", node, retries)
}

func CurrentEpoch(v Variables, node string) (int64, error) {
	items, _, err := v.Get(epochPath(node))
	if err != nil {
		return 0, err
	}
	return atoi(items["epoch"]), nil
}

// Clock is seconds on a monotonic origin; tests hand in their own.
type Clock func() float64

func Monotonic() Clock {
	start := time.Now()
	return func() float64 { return time.Since(start).Seconds() }
}

// Lease is held by the running instance. Renew is a read of the epoch
// Variable; MayWrite is a purely local decision on a monotonic clock.
type Lease struct {
	Vars        Variables
	Node        string
	Epoch       int64
	TTL, Margin float64
	Clock       Clock
	LastRenewal float64
	Fenced      bool
	Conflicts   int // node_epoch_conflicts
}

func NewLease(v Variables, node string, epoch int64, ttl, margin float64, clock Clock) *Lease {
	return &Lease{Vars: v, Node: node, Epoch: epoch, TTL: ttl, Margin: margin, Clock: clock, LastRenewal: clock()}
}

// Renew reports whether the lease still holds. False (and Fenced) if the
// cluster issued a newer epoch — or false, not fenced, if the read failed
// and the TTL ran out.
func (l *Lease) Renew() bool {
	if l.Fenced {
		return false
	}
	live, err := CurrentEpoch(l.Vars, l.Node)
	if err != nil { // the cluster is unreachable; keep going until TTL − margin
		return l.MayWrite()
	}
	if live != l.Epoch {
		l.Fenced = true
		l.Conflicts++
		return false
	}
	l.LastRenewal = l.Clock()
	return true
}

func (l *Lease) MayWrite() bool {
	if l.Fenced {
		return false
	}
	return l.Clock()-l.LastRenewal < l.TTL-l.Margin
}

func (l *Lease) SecondsLeft() float64 {
	left := (l.TTL - l.Margin) - (l.Clock() - l.LastRenewal)
	if left < 0 {
		return 0
	}
	return left
}

func atoi(s string) int64 {
	n, _ := strconv.ParseInt(s, 10, 64)
	return n
}
